numba/inline_closurecall.py

```diff
@@ -203,19 +246,435 @@ def _replace_returns(blocks, target, return_label):
     Return return statement by assigning directly to target, and a jump.
     """
     for label, block in blocks.items():
+        casts = []
         for i in range(len(block.body)):
             stmt = block.body[i]
             if isinstance(stmt, ir.Return):
                 assert(i + 1 == len(block.body))
                 block.body[i] = ir.Assign(stmt.value, target, stmt.loc)
                 block.body.append(ir.Jump(return_label, stmt.loc))
+                # remove cast of the returned value
+                for cast in casts:
+                    if cast.target.name == stmt.value.name:
+                        cast.value = cast.value.value
+            elif isinstance(stmt, ir.Assign) and isinstance(stmt.value, ir.Expr) and stmt.value.op == 'cast':
+                casts.append(stmt)
+
 
-def _add_definition(func_ir, block):
+def _add_definitions(func_ir, block):
     """
-    Add variable definitions to parent func_ir
+    Add variable definitions found in a block to parent func_ir.
     """
     definitions = func_ir._definitions
     assigns = block.find_insts(ir.Assign)
     for stmt in assigns:
         definitions[stmt.target.name].append(stmt.value)
 
+class GuardException(Exception):
+    pass
+
+def require(cond):
+    """
+    Raise GuardException if the given condition is False.
+    """
+    if not cond:
+       raise GuardException
+
+def guard(func, *args):
```

Comment:
I have to admit this was an in part academic interest question, I had a hack at the code to try and do it with a ctx manager and it shorten/cleaned some parts but I got stuck (potentially irretrievably) in a couple of places, so was interested in your view. Require-guard as an idiom for this seems as good a way as any in Python. I imagine if there was an `Optional` type cf. Java 8's `java.util.Optional` that help in another way, but again, different language :)